contracts/ERC20.sol

```diff
@@ -3,70 +3,242 @@ pragma solidity ^0.8.24;
+     * @param amount The amount of tokens to approve.
+     * @return True if the operation was successful.
+     * @dev Emits an {Approval} event.
+     */
     function approve(address spender, uint256 amount) external returns (bool) {
+        if (spender == address(0)) {
+            revert InvalidRecipient();
+        }
         allowance[msg.sender][spender] = amount;
         emit Approval(msg.sender, spender, amount);
         return true;
     }
 
-    function transferFrom(address sender, address recipient, uint256 amount)
-        external
-        returns (bool)
-    {
+    /**
+     * @notice Transfers `amount` tokens from `sender` to `recipient` using the allowance mechanism.
+     * @param sender The address of the sender.
+     * @param recipient The address of the recipient.
+     * @param amount The amount of tokens to transfer.
+     * @return True if the operation was successful.
+     * @dev Emits a {Transfer} event.
+     */
+    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool) {
         require(msg.sender != recipient, "cannot transfer to self");
-        allowance[sender][msg.sender] -= amount;
-        balanceOf[sender] -= amount;
-        balanceOf[recipient] += amount;
+        if (recipient == address(0)) {
+            revert InvalidRecipient();
+        }
+
+        uint b = allowance[sender][msg.sender];
+        uint c = b - amount;
+        if (c > b) {
+            revert InsufficientBalance();
+        }
+
+        unchecked {
+            allowance[sender][msg.sender] -= amount;
+            balanceOf[sender] -= amount;
+            balanceOf[recipient] += amount;
+        }
+
         emit Transfer(sender, recipient, amount);
         return true;
     }
 
+    /**
+     * @dev Internal function to mint `amount` tokens to `to`.
+     * @param to The address to receive the minted tokens.
+     * @param amount The amount of tokens to mint.
+     * @dev Emits a {Transfer} event from the zero address.
+     */
     function _mint(address to, uint256 amount) internal {
-        balanceOf[to] += amount;
-        totalSupply += amount;
+        if (to == address(0)) {
+            revert InvalidRecipient();
+        }
+
+        if (totalSupply + amount < totalSupply) {
+            revert Overflow();
+        }
+
+        unchecked {
+            balanceOf[to] += amount;
+            totalSupply += amount;
+        }
         emit Transfer(address(0), to, amount);
     }
 
+    /**
+     * @dev Internal function to burn `amount` tokens from `from`.
+     * @param from The address from which tokens will be burned.
+     * @param amount The amount of tokens to burn.
+     * @dev Emits a {Transfer} event to the zero address.
+     */
     function _burn(address from, uint256 amount) internal {
-        balanceOf[from] -= amount;
-        totalSupply -= amount;
+        if (totalSupply - amount > totalSupply) {
+            revert Underflow();
+        }
+
```

Comment:
Loop hole in function: It does not check whether the `from` address has enough tokens to be burnt from